bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/internal/item/ItemResource.java

```diff
@@ -590,29 +590,35 @@ public Response createOrUpdateItem(final @Context UriInfo uriInfo, final @Contex
         }
 
-        Item newItem = ItemDTOMapper.map(item, itemBuilderFactory);
-        if (newItem == null) {
-            logger.warn("Received HTTP PUT request at '{}' with an invalid item type '{}'.", uriInfo.getPath(),
-                    item.type);
-            return Response.status(Status.BAD_REQUEST).build();
-        }
+        try {
+            Item newItem = ItemDTOMapper.map(item, itemBuilderFactory);
+            if (newItem == null) {
+                logger.warn("Received HTTP PUT request at '{}' with an invalid item type '{}'.", uriInfo.getPath(),
+                        item.type);
+                return Response.status(Status.BAD_REQUEST).build();
+            }
 
-        // Save the item
-        if (getItem(itemname) == null) {
-            // item does not yet exist, create it
-            managedItemProvider.add(newItem);
-            return getItemResponse(uriBuilder(uriInfo, httpHeaders), Status.CREATED, itemRegistry.get(itemname), locale,
-                    null);
-        } else if (managedItemProvider.get(itemname) != null) {
-            // item already exists as a managed item, update it
-            managedItemProvider.update(newItem);
-            return getItemResponse(uriBuilder(uriInfo, httpHeaders), Status.OK, itemRegistry.get(itemname), locale,
-                    null);
-        } else {
-            // Item exists but cannot be updated
-            logger.warn("Cannot update existing item '{}', because is not managed.", itemname);
-            return JSONResponse.createErrorResponse(Status.METHOD_NOT_ALLOWED,
-                    "Cannot update non-managed Item " + itemname);
+            // Save the item
+            if (getItem(itemname) == null) {
+                // item does not yet exist, create it
+                managedItemProvider.add(newItem);
+                return getItemResponse(uriBuilder(uriInfo, httpHeaders), Status.CREATED, itemRegistry.get(itemname),
+                        locale, null);
+            } else if (managedItemProvider.get(itemname) != null) {
+                // item already exists as a managed item, update it
+                managedItemProvider.update(newItem);
+                return getItemResponse(uriBuilder(uriInfo, httpHeaders), Status.OK, itemRegistry.get(itemname), locale,
+                        null);
+            } else {
+                // Item exists but cannot be updated
+                logger.warn("Cannot update existing item '{}', because is not managed.", itemname);
+                return JSONResponse.createErrorResponse(Status.METHOD_NOT_ALLOWED,
+                        "Cannot update non-managed Item " + itemname);
+            }
+        } catch (IllegalArgumentException e) {
+            logger.warn("Received HTTP PUT request at '{}' with an invalid item name '{}'.", uriInfo.getPath(),
+                    item.name);
+            return Response.status(Status.BAD_REQUEST).build();
```

Comment:
The OpenAPI doc should be improved for this response since this response is currently unexpected for non-null item names:

```java
@ApiResponse(responseCode = "400", description = "Item null.")
```

Reply:
I also see that the method is annotated with:

```java
@Path("/{itemname: [a-zA-Z_0-9]+}")
```

Do you know if it returns the same response code for item names that do not match this expression?
It might be confusing if there are serveral response codes for invalid item names.

Reply:
It actually results in a 404.
And the topic is rather complex: The regex is checking for the path of the request - and our code only uses that path for determining if a new item should be created or an existing one is to be updated.
The request itself then contains the item as json payload - and if this is wrong, we (imho correctly) use a 400.

You can actually send the request to /items/ABC and have as payload XYZ as the item name - it will happily create the XYZ item for you... I have now added another check for that case as well (which will return HTTP 400, i.e. "invalid payload").